src/Microsoft.Extensions.DotnetToolDispatcher.Sources/DotnetToolDispatcher.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.PlatformAbstractions;
+
+namespace Microsoft.Extensions.Internal
+{
+    internal static class DotnetToolDispatcher
+    {
+        private const string DispatcherVersionArgumentName = "--dispatcher-version";
+
+        private static readonly string ToolName = PlatformServices.Default.Application.ApplicationName;
+        private static readonly string DispatcherVersionArgumentValue = ResolveDispatcherVersionArgumentValue();
```

Comment:
Also, can we override? We don't care if the assembly versions match directly as long as the "protocol" version matches, right? This "protocol" can be independent of the assembly versions.